web/server/web_client.c

```diff
@@ -812,109 +849,348 @@ static inline char *http_header_parse(struct web_client *w, char *s, int parse_u
+ */
+int web_client_parse_request(struct web_client *w,char *divisor) {
+    if(!divisor) {
+        w->total_params = 0;
+        return 0;
+    }
+
+    uint32_t i = url_parse_query_string(w->param_name,w->param_values,w->query_string.body+1,divisor);
+    w->total_params = i;
+
+    return i;
+}
+
+/**
+ *  Set Directory
+ *
+ *  Map directory variables
+ *
+ * @param w is the structure with all information of the client request.
+ * @param begin is the end of the first directory
+ * @param enddir a pointer to the position with the last valid character
+ * @param endcmd the end of the last character.
+ */
+static inline void web_client_set_directory(struct web_client *w,char *begin,char *enddir,char *endcmd) {
+    if (enddir) {
+        w->directory.body = begin;
+        w->directory.length = enddir - begin;
+
+        if (!strncmp(w->directory.body,"api",w->directory.length)) {
+            begin = enddir + 1;
+            enddir = strchr(begin,'/');
+            if(enddir) {
+                w->version.body = begin;
+                w->version.length = enddir - begin;
+
+                enddir++;
+                w->command.body = enddir;
+                w->command.length = (size_t) (endcmd - enddir);
+            }
+        }
+    }
+    else{
+        w->directory.body = begin;
+        w->directory.length = w->path.length - 1;
+        w->version.body = NULL;
+        w->version.length = 0;
+        w->command.body = NULL;
+        w->command.length = 0;
+    }
+}
+
+/**
+ *  Set without query string
+ *
+ *  When I do not have query string, it is necessary to set some default values.
+ *
+ * @param w is the structure with all information of the client request.
+ */
+static inline void web_client_set_without_query_string(struct web_client *w) {
+    w->query_string.body = NULL;
+    w->query_string.length = 0;
+
+    char *test = w->path.body+1;
+    if (!strncmp(test,"api/v1/",7) ) {
+        test += 7;
+        char *endofcommand=strchr(test,' ');
+        w->command.length=(endofcommand)?(size_t)endofcommand-(size_t)test:strlen(test);
+    } else {
+        w->command.length = w->path.length;
+    }
+    w->command.body = test;
+    w->total_params = 0;
+}
+
+/**
+ *  Split path and query
+ *
+ *  This function splits path and query string case there is one.
+ *
+ * @param w is the structure with all information of the client request.
+ */
+static inline void web_client_split_path_query(struct web_client *w) {
+    //I am working now with the decoded URL instead the original
+    w->path.body = w->decoded_url;
+    w->decoded_length = strlen(w->decoded_url);
+
+    //Is there a separator?
+    char *moveme = strchr(w->path.body,'?');
+    char *enddir;
+    if (moveme) { //I have a query string
+        w->path.length = moveme - w->path.body;
+        w->query_string.body = moveme;
+        w->query_string.length = w->decoded_length - w->path.length;
+
+        enddir = strchr(w->path.body+1,'/');
```

Comment:
Logic: can `strchr()` return NULL here? Can `w->path.body+1` be out of bounds?

Reply:
In the theory this would not happen, because we have before this a search by "?" that is the start of the query string, and for we have the interrogation we need to have before, ate least something like "/TEXT?", but you are correct, because in the practice somebody can send a bad request having only "?".